File: src/main/scala/Interpreter.scala
import parser.*
import ArithmaticOps.{Add, Div, Expo, Mod, Mul, Sub}
import BooleanOps.{And, Not, Or}
import CompareOps.{Eq, Greater, GreaterEq, Less, LessEq, NotEq}


val builtins = stdlib.stdlib

type HashMap[K, V] = scala.collection.mutable.HashMap[K, V]

class Scope(
    parent: Scope = null,
    funArgs: Seq[String] = Seq(),
    callArgs: Seq[Value] = Seq()
):
  private var parentScope: Scope = parent
  private var _result: Value = null
  private var _capture: Capture = null
  private var localVars: HashMap[String, Value] =
    new HashMap().addAll(funArgs.zip(callArgs))
  private var localFuncs: HashMap[String, parser.Function] = new HashMap

  def result: Option[Value] =
    if (this._result == null) None
    else
      val tmp = this._result
      this._result = null
      Some(tmp)

  def hasResult: Boolean = this._result != null
  def isGlobal: Boolean = this.parentScope == null

  def returnValue(value: Value): Scope =
    this._result = value
    this

  def withCapture(c: Capture): Scope =
    var tmp = this.clone()
    tmp._capture = c
    tmp

  private def lookupCapture(identifier: Identifier): Option[Value] =
    if (this._capture != null) this._capture.lookup(identifier)
    else None

  def lookup(identifier: Identifier): Option[Value] =
    this.lookupCapture(identifier) match {
      case Some(value) =>
        return Some(value)
      case None => {}
    }

    this.localVars.get(identifier.name) match {
      case Some(Identifier(name)) =>
        if (name == identifier.name) // 'x' -> 'x'
          lookupInParent(identifier)
        else
          this.lookup(Identifier(name))
      case Some(value) => Some(value)
      case None =>
        lookupInParent(identifier) match
          case Some(value) => Some(value)
          case None        => this.lookupFunction(identifier)

    }

  def lookupFunction(identifier: Identifier): Option[parser.Function] =
    this.localFuncs.get(identifier.name) match {
      case value: Some[parser.Function] => value
      case None                         => lookupFunctionInParent(identifier)
    }

  private def lookupInParent(identifier: Identifier): Option[Value] =
    if (this.parentScope != null)
      this.parentScope.lookup(identifier)
    else
      None

  private def lookupFunctionInParent(
      identifier: Identifier
  ): Option[parser.Function] =
    if (this.parentScope != null)
      this.parentScope.lookupFunction(identifier)
    else
      None

  override def clone(): Scope =
    var tmp = new Scope
    tmp.localFuncs = this.localFuncs.clone()
    tmp.localVars = this.localVars.clone()
    tmp.parentScope =
      this.parentScope // NOTE: intentional, parent scope is never modified
    tmp

  def update(identifier: Identifier, value: Value): Scope =
    value match {
      case f: Function =>
        this.lookupFunction(identifier) match {
          case None =>
            this.localFuncs.update(identifier.name, f)
          case Some(func: Function) =>
            f.capture = Some(Capture(this, f.args, f.body))
            this.localFuncs.update(identifier.name, f)
        }
      case v: Value =>
        this.localVars.update(identifier.name, v)
    }
    this
end Scope

class Capture(
    scope: Scope,
    functionArguments: Seq[String],
    functionBoby: Seq[Statement]
):
  private var captures: HashMap[String, Value] = {
    val externals = allExternalIdentifiers(functionArguments, functionBoby)
    val pairs = externals.map(e =>
      scope.lookup(Identifier(e)) match {
        case Some(value) =>
          val Some(v) = evalValue(value, scope).result: @unchecked
          (e, v)
        case None =>
          if (e == "print" || builtins.contains(e))
            ("_", Function(Seq(), Seq(Expression(Bool(false))), null))
          else
            assert(false, s"'$e' can not be captured because it does not exist")
      }
    )
    var tmp = new HashMap[String, Value]
    tmp.addAll(pairs)
  }

  private def allExternalIdentifiers(
      funcArgs: Seq[String],
      functionBoby: Seq[Statement]
  ): Seq[String] = {
    def externalsValue(
        accs: Seq[String],
        locals: Seq[String],
        value: Value
    ): (Seq[String], Seq[String]) =
      val cs = externalsOfValue(accs ++ locals, value)
        .flatMap(e => if (locals.contains(e)) Seq() else Seq(e))
      (cs, locals)

    def externalsStatements(
        accs: Seq[String],
        locals: Seq[String],
        sts: Seq[Statement]
    ): (Seq[String], Seq[String]) =
      val cs = allExternalIdentifiers(accs ++ locals, sts)
        .flatMap(e => if (locals.contains(e)) Seq() else Seq(e))
      (cs, locals)

    val (ext, _) =
      functionBoby.foldLeft((Seq(): Seq[String], funcArgs))((acc, elem) =>

        var (accs, locals) = acc
        elem match {
          case Assignment(varName, value) =>
            if (!accs.contains(varName) && !locals.contains(varName)) {
              // NOTE: very expensive, lots of copies here
              val (cs, ls) = externalsValue(accs, locals, value)
              (cs, ls.appended(varName))
            } else (accs, locals)
          case fc: FunctionCall =>
            externalsValue(accs, locals, fc)
          case Return(value) =>
            externalsValue(accs, locals, value)
          case WhileLoop(loop) =>
            val (accN, localsN) = externalsValue(accs, locals, loop.condition)
            externalsStatements(accN, localsN, loop.body)
          case Expression(expr) =>
            externalsValue(accs, locals, expr)
          case If(ifBranch, elifBranches, elseBranch) =>
            val varsIf = externalsValue(accs, locals, ifBranch.condition)
            val ifAcc = externalsStatements(varsIf._1, varsIf._2, ifBranch.body)

            val res: (Seq[String], Seq[String]) =
              elifBranches.foldLeft(ifAcc)((a, br) =>
                val (acc, loc) = a
                val (acc1, loc1) = externalsValue(acc, loc, br.condition)
                externalsStatements(acc1, loc1, br.body)
              )

            elseBranch match {
              case Some(br) =>
                val (accs, locals) = res
                externalsStatements(accs, locals, br)
              case None =>
                res
            }
        }
      )
    ext
  }

  private def externalsOfValue(
      currentCaptures: Seq[String],
      value: Value
  ): Seq[String] =
    value match {
      case Identifier(name) =>
        if (!currentCaptures.contains(name))
          currentCaptures.appended(name)
        else currentCaptures
      case Wrapped(value) => externalsOfValue(currentCaptures, value)
      case FunctionCall(functionExpr, args) =>
        val fx = externalsOfValue(currentCaptures, functionExpr)
        args.foldLeft(fx)(externalsOfValue)
      case Function(args, body, _capt) =>
        val caps = allExternalIdentifiers(args, body)
        caps.flatMap(e => if (args.contains(e)) Seq() else Seq(e))
      case ArrayLiteral(elements) =>
        elements.foldLeft(currentCaptures)(externalsOfValue)
      case _: Value => currentCaptures
    }

  def lookup(id: Identifier): Option[Value] =
    this.captures.get(id.name)
end Capture

def evalFunctionCall(
    functionExpr: Value,
    callArgs: Seq[Value],
    scope: Scope
): Scope =
  val evaledCallArgs = callArgs.map(evalValue(_, scope).result.get)
  functionExpr match {
    case Identifier(identifier) =>
      if (identifier == "print") {
        printValues(evaledCallArgs, scope)
        print("\n")
        scope
      } else if (builtins.contains(identifier)) {
        val callArgsNew = evaledCallArgs.map { value =>
          interpreterdata.toDataObject(value)
        }
        val Some(func) = builtins.get(identifier): @unchecked
        assert(
          func.n_args == callArgs.length,
          s"function call: expected ${func.n_args} argument(s) but got ${callArgs.length}"
        )
        val result = func.function(callArgsNew)
        scope.returnValue(interpreterdata.toAstNode(result))
      } else {
        scope.lookupFunction(Identifier(identifier)) match {
          case Some(Function(args, body, c)) =>
            // TODO: merge the capture with the current scope
            val res =
              body.foldLeft(
                Scope(scope.withCapture(c.getOrElse(null)), args, callArgs)
              )(
                evalStatement
              )
            val Some(value) = res.result: @unchecked
            scope.returnValue(value)
          case None =>
            assert(
              false,
              s"function '$identifier' not found"
            )
        }
      }
    case Wrapped(value) =>
      evalFunctionCall(value, evaledCallArgs, scope)

    case Function(args, body, s) =>
      val res =
        body.foldLeft(
          Scope(scope.withCapture(s.getOrElse(null)), args, evaledCallArgs)
        )(
          evalStatement
        )
      val Some(value) = res.result: @unchecked
      scope.returnValue(value)

    case FunctionCall(functionExpr, args) =>
      val Some(Function(args1, body1, s)) =
        evalFunctionCall(functionExpr, args, scope).result: @unchecked
      val res = body1.foldLeft(
        Scope(scope.withCapture(s.getOrElse(null)), args1, evaledCallArgs)
      )(evalStatement)
      val Some(value) = res.result: @unchecked
      scope.returnValue(value)
  }

def evalReturn(value: Value, scope: Scope): Scope =
  if (scope.isGlobal) assert(false, "can not return from global scope")
  else
    value match {
      case id: Identifier =>
        (scope.result, scope.lookup(id)) match {
          case (None, Some(v)) => scope.returnValue(v)
          case (Some(v), _)    => scope.returnValue(v)
          case (None, None) =>
            assert(
              false,
              s"identifier '${id.name}' does not exist that could be returned"
            )
        }

      case f: Function =>
        val tmp = Function(f.args, f.body, Some(Capture(scope, f.args, f.body)))
        scope.returnValue(f)
      case va =>
        scope.result match {
          case None =>
            val Some(res) = evalValue(va, scope).result: @unchecked
            scope.returnValue(res)
          case Some(v) => scope.returnValue(v)
        }
    }

def evalStatement(
    scope: Scope,
    st: Statement
): Scope =
  // First we differentiate of what type the current statement is.
  // The types are given by the `statementP` parsing rule in the Parser file.
  // Our goal of this function obviously is to handle all statements. This can either be a definition/re-definition
  // of a variable. Then we want to update the Scope. Or loops, if-else, function calls, return (see statementP rule).
  if (scope.hasResult)
    scope
  else
    st match {
      case Assignment(name, value) =>
        // Assignments have the form variable = Value. See the `valueP` rule of the parser for the types that are assignable.
        // Our goal here is it to evaluate the right-hand side. Eg. x = 4+9, then we want to add the variable x to the
        // Scope with the value 13 as a number.
        val result = evalValue(value, scope)
        // NOTE: `.result` should allways exist. It indicates a bug in `evalValue` if it does not
        // The evalValue-function stores the newly calculated value in the Scope in the variable `result`. We need to
        // update the variable `name` with this new value.
        val Some(newValue) = result.result: @unchecked
        scope.update(Identifier(name), newValue)
      case FunctionCall(identifier, callArgs) =>
        evalFunctionCall(identifier, callArgs, scope)
      case Return(value) =>
        evalReturn(value, scope)
      case Expression(expr) =>
        val Some(result) = evalValue(expr, scope).result: @unchecked
        scope.returnValue(result)
      case If(ifBranch, elifBranches, elseBranch) => {
        evalIf(ifBranch, elifBranches, elseBranch, scope)
      }
      case WhileLoop(loop) =>
        evalWhileLoop(loop, scope)
    }

def evalWhileLoop(whileLoop: Branch, scope: Scope): Scope = {
  var currentScope = scope
  while (
    evalValue(whileLoop.condition, currentScope).result.contains(Bool(true))
  ) {
    currentScope =
      whileLoop.body.foldLeft(currentScope)((accScope, statement) => {
        if (accScope.hasResult)
          accScope // Early exit if result is set (like return statements)
        else evalStatement(accScope, statement)
      })
  }
  currentScope
}

def evalIf(
    initialBranch: Branch,
    elifBranches: Seq[Branch],
    elseBranch: Option[Seq[Statement]],
    scope: Scope
): Scope = {

  def evalBranch(branch: Branch, scope: Scope): Option[Scope] = {
    val result = evalValue(branch.condition, scope)
    result.result match {
      case Some(Bool(true)) =>
        Some(branch.body.foldLeft(scope)(evalStatement))
      case _ =>
        None
    }
  }

  // Check initial branch condition
  var conditionsMet = false
  evalBranch(initialBranch, scope) match {
    case Some(updatedScope) =>
      updatedScope // Return updated scope if the initial branch is true
    case None =>
      val finalScope = elifBranches.foldLeft(scope) { (currentScope, branch) =>
        evalBranch(branch, currentScope) match {
          case Some(updatedScope) =>
            conditionsMet = true
            updatedScope
          case None => currentScope
        }
      }

      if (conditionsMet) {
        return finalScope
      } else if (elseBranch.isEmpty) {
        return scope
      }

      // Evaluate the else branch if no conditions were met
      elseBranch match {
        case Some(statements) =>
          statements.foldLeft(finalScope)(evalStatement)
        case _ => finalScope
      }
  }
}

def evalValue(
    v: Value,
    scope: Scope
): Scope =
  v match {
    case Function(args, body, _) =>
      scope.returnValue(Function(args, body, Some(Capture(scope, args, body))))
    case FunctionCall(identifier, callArgs) =>
      evalFunctionCall(identifier, callArgs, scope)
    case BinaryOp(left, op, right) =>
      val left_result = evalValue(left, scope)
      val Some(new_left) = left_result.result: @unchecked
      val right_result = evalValue(right, scope)
      val Some(new_right) = right_result.result: @unchecked
      evalBinaryOp(op, new_left, new_right, scope)
    case UnaryOp(op, value) =>
      val Some(result) = evalValue(value, scope).result: @unchecked
      op match {
        case ArithmaticOp(Add) => scope.returnValue(result)
        case ArithmaticOp(Sub) =>
          result match {
            case n: Number =>
              scope.returnValue(Number(-n.value))
            case v =>
              assert(false, s"unary op: value case '$v' is not implemented")
          }
        case BooleanOp(Not) =>
          val tmp = result match {
            case Bool(value) => Bool(!value)
            case v =>
              assert(false, s"unary operator 'not' is not defined for '$v'")
          }
          scope.returnValue(tmp)
        case o => assert(false, s"unary op: op case '$o' is not implemented")
      }
    case StructureAccess(id, v) => {
      val returnVal: Value = scope.lookup(id) match {
        case Some(Dictionary(entries)) =>
          val result: Option[Value] = entries.foldLeft(None) { (acc, curr) =>
            acc match {
              case None =>
                val res = evalValue(curr.key, scope)
                val Some(value) = res.result: @unchecked
                val res2 = evalCompareOps(Eq, value, v, scope)
                res2.result match {
                  case Some(Bool(true)) =>
                    Some(curr.value)
                  case _ => None
                }
              case r => r
            }
          }
          result match {
            case Some(value) => value
            case None => assert(false, s"structure '${id.name}' does not exist")
          }
        case Some(ArrayLiteral(entries)) => {
          evalValue(v, scope).result match {
            case None => assert(false, s"Expr \"$v\" is not interpretable")
            case Some(Number(n)) => {
              if (n != n.toInt) {
                throw IllegalArgumentException("expected hole number, but got number with decimal part")
              }  
              entries(n.toInt)
            }
            case x => throw IllegalArgumentException("expected number, not: " + x.toString)
          }
        }
        case _ =>
          assert(false, s"no structure found by the name '${id.name}'")
      }
      scope.returnValue(returnVal)
    }
    case Identifier(name) =>
      scope.lookup(Identifier(name)) match {
        case Some(value) =>
          scope.returnValue(value)
        case None =>
          assert(false, s"identifier '$name' does not exist")
      }
    case Number(value) =>
      scope.returnValue(Number(value))
    case StdString(value) =>
      scope.returnValue(StdString(value))
    case Bool(value) =>
      scope.returnValue(Bool(value))
    case Wrapped(value) =>
      evalValue(value, scope)
    case Dictionary(entries) =>
      scope.returnValue(Dictionary(entries))
    case ArrayLiteral(elements) => 
      scope.returnValue(ArrayLiteral(elements))
    case err =>
      assert(false, f"TODO: not implemented '$err'")
  }

def evalBinaryOp(
    op: Operator,
    left: Value,
    right: Value,
    scope: Scope
): Scope = {
  op match
    case ArithmaticOp(ops) => evalArithmeticOps(ops, left, right, scope)
    case CompareOp(ops)    => evalCompareOps(ops, left, right, scope)
    case BooleanOp(ops)    => evalBooleanOps(ops, left, right, scope)
    case _ => throw new IllegalArgumentException("Binary operation invalid.")
}

def evalBooleanOps(
    op: BooleanOps,
    left: Value,
    right: Value,
    scope: Scope
): Scope = {
  // evaluate left and right value
  val Some(leftEval) = evalValue(left, scope).result: @unchecked
  val Some(rightEval) = evalValue(right, scope).result: @unchecked

  // if left or right is string.
  if (leftEval.isInstanceOf[Identifier] || rightEval.isInstanceOf[Identifier]) {
    assert(false, "No boolean operators allowed on strings")
  }

  // otherwise left and right are bools or numbers
  val result = op match {
    case And => (extractNumber(leftEval) > 0) && (extractNumber(rightEval) > 0)
    case Or  => (extractNumber(leftEval) > 0) || (extractNumber(rightEval) > 0)
    case _   => assert(false, "Unexpected comparison operation.")
  }

  scope.returnValue(Bool(result)) // Adding the result to the scope
}

def evalCompareOps(
    op: CompareOps,
    left: Value,
    right: Value,
    scope: Scope
): Scope = {
  // evaluate left and right value
  val Some(leftEval) = evalValue(left, scope).result: @unchecked
  val Some(rightEval) = evalValue(right, scope).result: @unchecked
  // if left or right is string.
  (leftEval, rightEval) match {
    case (s1: StdString, s2: StdString) => {
      op match {
        case Eq =>
          scope.returnValue(Bool(s1.value == s2.value))
        case NotEq =>
          scope.returnValue(Bool(s1.value != s2.value))
        case _ =>
          assert(
            false,
            "For Strings only the operators '==' and '!=' are allowed."
          )
      }
    }
    case (value1: (Number | Bool), value2: (Number | Bool)) => {
      // otherwise left and right are bools or numbers
      val result = op match {
        case Less      => extractNumber(value1) < extractNumber(value2)
        case LessEq    => extractNumber(value1) <= extractNumber(value2)
        case Greater   => extractNumber(value1) > extractNumber(value2)
        case GreaterEq => extractNumber(value1) >= extractNumber(value2)
        case Eq        => value1 == value2
        case NotEq     => value1 != value2
      }

      scope.returnValue(Bool(result)) // Adding the result to the scope
    }
    case (v1, v2) =>
      assert(
        false,
        s"the values '$v1' and '$v2' are not comparable under '$op'"
      )
  }
}

def typeOf(value: Value): String =
  value match {
    case _: StdString  => "string"
    case _: Number     => "number"
    case _: Bool       => "bool"
    case _: Dictionary => "dictionary"
    // case _: ArrayLiteral => "array"
    case _: Function => "function"
    case _           => "'not defined'"
  }

def evalArithmeticOps(
    op: ArithmaticOps,
    left: Value,
    right: Value,
    scope: Scope
): Scope = {
  // evaluate left and right value
  val Some(leftEval) = evalValue(left, scope).result: @unchecked
  val Some(rightEval) = evalValue(right, scope).result: @unchecked

  (leftEval, rightEval) match {
    case (s1: StdString, s2: StdString) =>
      op match {
        case Add =>
          val tmp = s1.value + s2.value
          scope.returnValue(StdString(tmp))
        case _ => assert(false, "only operator '+' is defined for strings")
      }
    case (value1: (Number | Bool), value2: (Number | Bool)) =>
      val leftNumber = extractNumber(value1)
      val rightNumber = extractNumber(value2)
      // calculate arithmetic operation
      val result = op match {
        case Add  => leftNumber + rightNumber
        case Sub  => leftNumber - rightNumber
        case Mul  => leftNumber * rightNumber
        case Div  => leftNumber / rightNumber
        case Mod  => leftNumber % rightNumber
        case Expo => scala.math.pow(leftNumber, rightNumber)
        case null => assert(false, "unreachable")
      }
      scope.returnValue(Number(result)) // Adding the result to the scope
    case (v1, v2) =>
      val type1 = typeOf(v1)
      val type2 = typeOf(v2)
      assert(false, s"Operator '$op' is not defined for '$type1' and '$type2'")
  }
}

// Input: Number or Bool. Output: Double (true == 1, false == 0)
def extractNumber(value: Value): Double = value match {
  case Number(n) => n
  case Bool(b)   => if (b) 1.0 else 0.0
  case v => assert(false, s"Expected number or boolean in comparison. Got '$v'")
}

def printValues(values: Seq[Value], scope: Scope): Unit =
  val output = values
    .map(_.toString)
    .mkString(" ")
  print(output)

File: src/zig/interpreter.zig
const std = @import("std");
const stmt = @import("statement.zig");
const expr = @import("expression.zig");

const Scope = @import("scope.zig");

const Expression = expr.Expression;
const Statement = stmt.Statement;

pub const Error = error{
    NotImplemented,
    FunctionNotFound,
    ValueNotFound,
    IOWrite,
    InvalidExpressoinType,
} || Scope.Error;

pub fn evalStatement(statement: Statement, scope: *Scope) Error!void {
    if (scope.hasResult() and !scope.isGlobal()) {
        return;
    }

    switch (statement) {
        .assignment => |a| {
            try evalExpression(a.value, scope);
            try scope.update(a.varName, scope.result() orelse unreachable);
        },
        .functioncall => |fc| {
            try evalFunctionCall(fc, scope);
            scope.return_result = null;
        },
        .ret => |r| {
            try evalExpression(r.value, scope);
            const result = scope.result() orelse unreachable;
            scope.return_result = result;
        },
        .whileloop => |w| {
            var cond = false;
            try evalExpression(w.loop.condition, scope);
            var tmp = scope.result() orelse unreachable;
            cond = tmp.* == .boolean and tmp.boolean.value;
            while (cond) {
                for (w.loop.body) |st| {
                    try evalStatement(st, scope);
                }

                try evalExpression(w.loop.condition, scope);
                tmp = scope.result() orelse unreachable;
                cond = tmp.* == .boolean and tmp.boolean.value;
            }
        },
        .if_statement => |i| {
            try evalExpression(i.ifBranch.condition, scope);
            const tmp = scope.result() orelse unreachable;
            if (tmp.* == .boolean and tmp.boolean.value) {
                for (i.ifBranch.body) |st| {
                    try evalStatement(st, scope);
                }
            } else {
                if (i.elseBranch) |b| {
                    for (b) |st| {
                        try evalStatement(st, scope);
                    }
                }
            }
        },
        else => |st| {
            std.debug.print("TODO: unhandled case in eval statement: {}\n", .{st});
            return Error.NotImplemented;
        },
    }
}

fn evalFunctionCall(fc: expr.FunctionCall, scope: *Scope) Error!void {
    const tmpArgs = try scope.allocator.alloc(Expression, fc.args.len);
    defer scope.allocator.free(tmpArgs);

    for (fc.args, tmpArgs) |*arg, *tmparg| {
        try evalExpression(arg, scope);
        const tmp = scope.result() orelse unreachable;
        tmparg.* = tmp.*;
    }

    switch (fc.func.*) {
        .identifier => |id| {
            if (std.mem.eql(u8, id.name, "print")) {
                var has_printed = false;
                for (tmpArgs) |*value| {
                    if (has_printed) {
                        scope.out.print(" ", .{}) catch return Error.IOWrite;
                    } else has_printed = true;
                    try printValue(value.*, scope);
                }
                scope.out.print("\n", .{}) catch return Error.IOWrite;
            } else if (false) { // stdlib functions
                return Error.NotImplemented;
            } else if (scope.lookupFunction(id)) |f| {
                var localScope = try Scope.init(scope.allocator, scope.out, scope, f.args, tmpArgs);
                for (f.body) |st| {
                    try evalStatement(st, &localScope);
                }
                const result = localScope.result();
                scope.return_result = result;
            }
        },
        else => |e| {
            std.debug.print("ERROR: unhandled case in function call: {}\n", .{e});
            return Error.NotImplemented;
        },
    }
}

fn printValue(value: Expression, scope: *Scope) Error!void {
    switch (value) {
        .number => |n| {
            if (n == .float) {
                scope.out.print("{}", .{n.float}) catch return Error.IOWrite;
            } else scope.out.print("{}", .{n.integer}) catch return Error.IOWrite;
        },
        .boolean => |v| {
            scope.out.print("{}", .{v.value}) catch return Error.IOWrite;
        },
        .string => |v| {
            scope.out.print("{s}", .{v.value}) catch return Error.IOWrite;
        },
        else => |v| {
            std.debug.print("TODO: printing of value: {}\n", .{v});
            return Error.NotImplemented;
        },
    }
}

fn evalExpression(value: *Expression, scope: *Scope) Error!void {
    switch (value.*) {
        .identifier => |id| {
            var v = try scope.lookup(id) orelse {
                std.debug.print("ERROR: no value found for '{s}'\n", .{id.name});
                return Error.ValueNotFound;
            };
            while (v.* == .identifier) {
                v = try scope.lookup(id) orelse {
                    std.debug.print("ERROR: no value found for '{s}'\n", .{id.name});
                    return Error.ValueNotFound;
                };
            }
            scope.return_result = v;
        },
        .binary_op => |bin| try evalBinaryOp(bin.op, bin.left, bin.right, scope),
        .unary_op => |un| try evalUnaryOp(un.op, un.operant, scope),
        .wrapped => |w| {
            try evalExpression(w, scope);
        },
        .functioncall => |fc| {
            try evalFunctionCall(fc, scope);
        },
        .array => {
            scope.return_result = value;
        },
        .dictionary => {
            scope.return_result = value;
        },
        .function => {
            scope.return_result = value;
        },
        .number => {
            scope.return_result = value;
        },
        .string => {
            scope.return_result = value;
        },
        .boolean => {
            scope.return_result = value;
        },
        else => |v| {
            std.debug.print("TODO: unhandled case in eval expr: {}\n", .{v});
            return Error.NotImplemented;
        },
    }
}

fn evalUnaryOp(op: expr.Operator, operant: *Expression, scope: *Scope) !void {
    switch (op) {
        .arithmetic => |ops| {
            if (ops != .Sub) unreachable;
            switch (operant.*) {
                .number => |num| {
                    if (num == .float) {
                        const tmp = try expr.Number.init(scope.allocator, f64, -num.float);
                        scope.return_result = tmp;
                    } else {
                        const tmp = try expr.Number.init(scope.allocator, i64, -num.integer);
                        scope.return_result = tmp;
                    }
                },
                else => return Error.NotImplemented,
            }
        },
        .boolean => |ops| {
            if (ops != .Not) unreachable;
            switch (operant.*) {
                .boolean => |b| {
                    const tmp = try expr.Boolean.init(scope.allocator, !b.value);
                    scope.return_result = tmp;
                },
                else => return Error.NotImplemented,
            }
        },
        else => unreachable,
    }
}

fn evalBinaryOp(op: expr.Operator, left: *Expression, right: *Expression, scope: *Scope) !void {
    switch (op) {
        .arithmetic => |ops| try evalArithmeticOps(ops, left, right, scope),
        .compare => |ops| try evalCompareOps(ops, left, right, scope),
        .boolean => |ops| try evalBooleanOps(ops, left, right, scope),
    }
}

fn evalArithmeticOps(op: expr.ArithmeticOps, left: *Expression, right: *Expression, scope: *Scope) !void {
    try evalExpression(left, scope);
    const leftEval = scope.result() orelse unreachable;
    try evalExpression(right, scope);
    const rightEval = scope.result() orelse unreachable;

    switch (op) {
        .Add => switch (leftEval.*) {
            .string => |l| switch (rightEval.*) {
                .string => |r| {
                    const out = try scope.allocator.create(Expression);
                    const tmp = try std.mem.join(scope.allocator, "", &[_][]const u8{ l.value, r.value });
                    out.* = .{ .string = .{ .value = tmp } };
                    scope.return_result = out;
                },
                else => return Error.NotImplemented,
            },
            .number => |l| switch (rightEval.*) {
                .number => |r| {
                    const n = l.add(r);
                    if (n == .float) {
                        const tmp = try expr.Number.init(scope.allocator, f64, n.float);
                        scope.return_result = tmp;
                    } else {
                        const tmp = try expr.Number.init(scope.allocator, i64, n.integer);
                        scope.return_result = tmp;
                    }
                },
                else => return Error.NotImplemented,
            },
            else => {
                std.debug.print("ERROR: can not add value of type '{}'\n", .{leftEval});
                return Error.NotImplemented;
            },
        },
        else => |v| {
            std.debug.print("ERROR: unhandled case in arith. bin. op.: {}\n", .{v});
            return Error.NotImplemented;
        },
    }
}

fn asNumber(ex: *Expression) expr.Number {
    return switch (ex.*) {
        .number => ex.number,
        .boolean => |bl| expr.Number{ .integer = @intFromBool(bl.value) },
        else => {
            std.debug.print("ERROR: value not convertable to number: {}\n", .{ex});
            unreachable;
        },
    };
}

fn evalCompareOps(op: expr.CompareOps, left: *Expression, right: *Expression, scope: *Scope) !void {
    try evalExpression(left, scope);
    const leftEval = if (scope.result()) |tmp| asNumber(tmp) else unreachable;
    try evalExpression(right, scope);
    const rightEval = if (scope.result()) |tmp| asNumber(tmp) else unreachable;

    switch (op) {
        .Equal => {
            const tmp = try scope.allocator.create(Expression);
            tmp.* = .{ .boolean = .{ .value = leftEval.eql(rightEval) } };
            scope.return_result = tmp;
        },
        .NotEqual => {
            const tmp = try scope.allocator.create(Expression);
            tmp.* = .{ .boolean = .{ .value = !leftEval.eql(rightEval) } };
            scope.return_result = tmp;
        },
        .Less => {
            const out = try scope.allocator.create(Expression);
            const tmp: bool = switch (leftEval) {
                .float => |l| switch (rightEval) {
                    .float => |r| l < r,
                    .integer => |r| l < @as(f64, @floatFromInt(r)),
                },
                .integer => |l| switch (rightEval) {
                    .float => |r| @as(f64, @floatFromInt(l)) < r,
                    .integer => |r| l < r,
                },
            };
            out.* = .{ .boolean = .{ .value = tmp } };
            scope.return_result = out;
        },
        .LessEqual => {
            const out = try scope.allocator.create(Expression);
            const tmp: bool = switch (leftEval) {
                .float => |l| switch (rightEval) {
                    .float => |r| l <= r,
                    .integer => |r| l <= @as(f64, @floatFromInt(r)),
                },
                .integer => |l| switch (rightEval) {
                    .float => |r| @as(f64, @floatFromInt(l)) <= r,
                    .integer => |r| l <= r,
                },
            };
            out.* = .{ .boolean = .{ .value = tmp } };
            scope.return_result = out;
        },
        else => return Error.NotImplemented,
    }
}

fn evalBooleanOps(op: expr.BooleanOps, left: *Expression, right: *Expression, scope: *Scope) !void {
    try evalExpression(left, scope);
    const leftEval = scope.result() orelse unreachable;
    try evalExpression(right, scope);
    const rightEval = scope.result() orelse unreachable;

    if (leftEval.* != .boolean or rightEval.* != .boolean) {
        std.debug.print("ERROR: boolean operators are only allowed for booleans\n", .{});
        return Error.NotImplemented;
    }

    const l = leftEval.boolean.value;
    const r = rightEval.boolean.value;

    switch (op) {
        .And => {
            const out = try scope.allocator.create(Expression);
            out.* = .{ .boolean = .{ .value = l and r } };
            scope.return_result = out;
        },
        .Or => {
            const out = try scope.allocator.create(Expression);
            out.* = .{ .boolean = .{ .value = l or r } };
            scope.return_result = out;
        },
        else => {
            unreachable;
        },
    }
}
